File: leetcode/C/multithread_Lock.c
#include <pthread.h>
// linux环境下使用 man pthread_mutex_init 查看帮助文档

typedef struct {
    int n;
    pthread_mutex_t foo_mutex;
    pthread_mutex_t bar_mutex;
} FooBar;

extern void printFoo();
extern void printBar();

FooBar* fooBarCreate(int n) {
    FooBar* obj = (FooBar*) malloc(sizeof(FooBar));
    obj->n = n;
    pthread_mutex_init(&obj->foo_mutex, NULL);
    pthread_mutex_init(&obj->bar_mutex, NULL);
    return obj;
}

void foo(FooBar* obj) {
    
    for (int i = 0; i < obj->n; i++) {
        // printFoo() outputs "foo". Do not change or remove this line.
        pthread_mutex_lock(&obj->foo_mutex);
        printFoo();
        pthread_mutex_unlock(&obj->bar_mutex);
    }
}

void bar(FooBar* obj) {
    
    for (int i = 0; i < obj->n; i++) {
        pthread_mutex_lock(&obj->bar_mutex);
        // printBar() outputs "bar". Do not change or remove this line.
        printBar();
        pthread_mutex_unlock(&obj->foo_mutex);
    }
}

void fooBarFree(FooBar* obj) {
    pthread_mutex_destroy(&obj->foo_mutex);
    pthread_mutex_destroy(&obj->bar_mutex);
}

File: leetcode/python/multithread_Lock.py
import threading

class FooBar(object):
    def __init__(self, n):
        self.n = n
        self.locks = {"foo": threading.Lock(), "bar": threading.Lock()}
        self.locks["bar"].acquire()

    def foo(self, printFoo):
        """
        :type printFoo: method
        :rtype: void
        """
        for i in range(self.n):
            self.locks["foo"].acquire()            
            # printFoo() outputs "foo". Do not change or remove this line.        	    
            printFoo()
            self.locks["bar"].release()

    def bar(self, printBar):
        """
        :type printBar: method
        :rtype: void
        """
        for i in range(self.n):
            self.locks["bar"].acquire()
            # printBar() outputs "bar". Do not change or remove this line.
            printBar()
            self.locks["foo"].release()

def printFoo():
    print("foo")

def printBar():
    print("bar")

if __name__ == "__main__":
    size = 10
    foobar = FooBar(size)
    threads = []
    for i in range(size):            
        threads.append(threading.Thread(target=foobar.foo, args=(printFoo,)))
        threads.append(threading.Thread(target=foobar.bar, args=(printBar,)))

    for i in range(size):
        threads[i].start()
